### PR TITLE
Restrict implicit conversions

### LINKED ISSUE #70: Implicit conversion through a constructor relying on SFINAE confuses the compiler
`std::is_convertible` over estimates the convertibility of types into dual numbers. This causes compile errors with Eigen, when instantiating Eigen objects on DualNumbers. The fix may be to restrict the 

```
template <typename T2>
DualNumber(const T2& val);
```

with a dash of `enable_if`.

### PR DESCRIPTION
**Pull request by dschwen**
Closes #70

### CODE REVIEW

**Comment by roystgnr** [comment]
I'd totally forgotten about @lindsayad 's NotADuckDualNumber stuff.  Looks like that using declaration is inadequate when the constructor we're trying to use has been overloaded?  Let me see if I can replicate this.

**Comment by dschwen** [comment]
Trying to catch up here. Was that not a duck number created for the outer-type-is-dualnumber approach? Because we're not doing that anywhere in MOOSE.

**Comment by lindsayad** [comment]
From https://en.cppreference.com/w/cpp/language/using_declaration, there is this passage:
> The semantics of inheriting constructors were retroactively changed by a defect report against C++11. Previously, an inheriting constructor declaration caused a set of synthesized constructor declarations to be injected into the derived class, which caused redundant argument copies/moves, had problematic interactions with some forms of SFINAE, and in some cases can be unimplementable on major ABIs. Older compilers may still implement the previous semantics. 
So I don't think that `using Base::Base` is even supported anymore from what I understand...

**Comment by lindsayad** [comment]
> Looks like that using declaration is inadequate when the constructor we're trying to use has been overloaded?

We had overloaded constructors before. I think it's the SFINAE bit that's choking the compiler. So yea I'm totally good with shimming instead of `using`

**Comment by lindsayad** [comment]
> Because we're not doing that anywhere in MOOSE.

@dschwen MOOSE is the world :smile:

**Comment by lindsayad** [comment]
I can see that this is a bigger problem. The general `NotADuckDualNumber` template was rarely intended to be used. Instead the user was meant to provide specializations for their special non-POD structures. The new error is out of one of these specializations that is also trying `using DualNumber.*::DualNumber`. So if someone is using `NotADuckDualNumber` for non-POD (which is all that it was initially intended to be used for) then this PR will likely break their code, e.g. it will break their code if they're trying to inherit `DualNumber` constructors

**Comment by lindsayad** [comment]
Our tests here can be fixed up and we will be able to pass CI, but this is a breaking change. My honest opinion is that the `NotADuckDualNumber` code should be removed. I believe that at one point in time I did some pen-to-paper work and realized that it's fairly easy to mess up chain-rule derivatives (or maybe chain-rule derivatives are in fact the thing messed up as is?) with the `NotADuckDualNumber`. My vote would be to eventually merge this PR with deletion of the `NotADuckDualNumber` code and mark as a new major release

**Comment by roystgnr** [comment]
> It's my world

Have you ever seen the Twilight Zone episode, "Button, Button"?  Beware of depending on libraries whose maintainers are indifferent to the fate of other users *who you don't even know* ...

> The general NotADuckDualNumber template was rarely intended to be used. Instead the user was meant to provide specializations for their special non-POD structures. The new error is out of one of these specializations

Okay, so forget about replacing ```using``` with shims, then, unless we force every user to create matching shims.

On the other hand, IMHO we have to do *something*.  I feel like a jackass after my reference above, but getting correct behavior out of ```is_convertible``` is actually important enough that we can require users to upgrade their codes to match.  I'd really like to avoid that if possible, though.  I just don't see how it's possible.  I'm hoping the DualNumber/ShadowNumber interaction might give me some clue but so far no luck.

**Comment by roystgnr** [comment]
> or maybe chain-rule derivatives are in fact the thing messed up as is?

And it turns out I actually *can* be talked into completely breaking some users' code, after all!  I'm not a fan of "your code won't compile", but it's a **huge** improvement over "your code will compile but can silently give you incorrect results", so if you're not kidding about the risk there then let's take it out behind the barn and shoot it.

**Comment by lindsayad** [comment]
K I'm looking more into our MOOSE history like idaholab/moose#12365. I'd forgotten that at one time we really did have the `NotADuckDualNumber` template as the core template in MOOSE. I'd thought that I'd made the switch back to `DualNumber` in the initial MOOSE AD PR (idaholab/moose#12111). The reasons I list in idaholab/moose#12365 don't say anything about wrongness so maybe I had fixed that.

**Comment by dschwen** [comment]
Ha, well, turns out , this is not fixing the problem at all. I'm getting an ambiguous conversion explicitly because of teh two constructors I added:

```
../src/numerics/include/metaphysicl/dualnumber_decl.h:67:3: note: candidate constructor [with T2 = MetaPhysicL::DualNumber<double, MetaPhysicL::DynamicSparseNumberVector<double, unsigned int>, false>]
  DualNumber(const T2& val, typename std::enable_if<std::is_convertible<typename T2::value_type,T>::value && std::is_convertible<typename T2::derivatives_type,D>::value, void*>::type = nullptr);
  ^
../src/numerics/include/metaphysicl/dualnumber_decl.h:70:3: note: candidate constructor [with T2 = MetaPhysicL::DualNumber<double, MetaPhysicL::DynamicSparseNumberVector<double, unsigned int>, false>]
  DualNumber(const T2& val, typename std::enable_if<std::is_convertible<T2,T>::value, void*>::type = nullptr);
  ^
```

If constexprs short circuit maybe I can fold them into one single constructor...

**Comment by roystgnr** [comment]
@dschwen on Slack last night: "Urgent might be an overstatement"

@dschwen this afternoon: "Do it."

doit

That ambiguity issue might be the same problem that ShadowNumber hit.  I'll push the fix to this branch shortly.

**Comment by dschwen** [comment]
Yeah, the "short circuit" idea does not work at all

```
../src/numerics/include/metaphysicl/dualnumber_decl.h:67:3: note: candidate template ignored: substitution failure [with T2 = float]: type 'float' cannot be used prior to '::' because it has no members
  DualNumber(const T2& val, typename std::enable_if<std::is_convertible<T2,T>::value || (std::is_convertible<typename T2::value_type,T>::value && std::is_convertible<typename T2::derivatives_type,D>::value), void*>::type = nullptr);
```

ah, maybe I can use enable if differently...

**Comment by roystgnr** [comment]
If we're not needing to delete NotADuckDualNumber after all, mind if I just force push my version of the branch here?  It uses the shim functions, which is ugly, but it passes internal tests now, which is nice...

**Comment by roystgnr** [comment]
Or you could try it out yourself first: https://github.com/roystgnr/MetaPhysicL/tree/is_convertible_70_roy

**Comment by dschwen** [comment]
feel free to use the force

**Comment by roystgnr** [comment]
Well, the tests all passed on *my* system, anyway ... :cry:

**Comment by dschwen** [comment]
Yeah, the hash you pushed looks good to me, and my Eigen test still builds fine with it!

**Comment by dschwen** [comment]
Looks like the ducks get to see another sunrise...

**Comment by roystgnr** [comment]
I'll let @lindsayad have the final word here but I think we're good to merge.

**Comment by roystgnr** [comment]
If the Eigen integration continues to work out, we should get it added to our test suite.  There's multiple other classes here which probably need the same treatment, but I don't trust myself to do it blindly, I want a clear failing test I can make work.  I also think it would be awesome to find out what ShadowNumber calculations say about FP roundoff error in complicated Eigen use cases.

**Comment by lindsayad** [comment]
My memory is so short. I appear to have fixed the chain rule stuff in https://github.com/roystgnr/MetaPhysicL/pull/22. I feel like I was smarter back then

File: src/numerics/include/metaphysicl/nddualnumber.h
Comment (roystgnr) [reply src/numerics/include/metaphysicl/nddualnumber.h:22]: Ah, hell, I didn't propagate my DualNumber fixes to NotADuckDualNumber, did I?  I'll do that now.  Good catch.

Comment (lindsayad) [reply src/numerics/include/metaphysicl/nddualnumber.h]: No I'm being dense. The top overload does `std::convertible<typename T2::value_type,T>` where the bottom overload does `std::is_convertible<T2,T>` so I don't really see a situation where we're going to get a type `T2` who has a `value_type` member that is convertible to `T` while also itself being implicitly convertible to `T`

**Review by lindsayad — APPROVED** [review]
I'm glad the non-ducks (eagles maybe?) get to see another sunrise

**Comment by lindsayad** [comment]
@roystgnr what problems were you getting with `ShadowNumber`? I'm looking at the `DualNumber` constructors, even before you removed the `value_type` SFINAE, and I don't understand under what circumstances there would have been ambiguity. I understand that `ShadowNumber` defines `value_type` but still

**Comment by roystgnr** [comment]
But that was all it took!  When the input type defined value_type but also was convertible to T (because it was a T, in this case, IIRC), then it qualified for both the first constructor and for the second.

**Comment by roystgnr** [comment]
Even nested DualNumbers might have hit the same weirdness, I think.  But now that the first constructor is a partial template specialization that's enough for the compiler to recognize that it takes priority over a more general template match.

**Comment by roystgnr** [comment]
Hmm.. I hope, anyway.  Now I'm worrying about whether we have enough DualNumber-of-DualNumber test coverage in MetaPhysicL itself.  I remember relying on that pretty heavily for https://manufactured-solutions.github.io/MASA/ but I never used it in libMesh and I don't know if you ever did in MOOSE.

... I'm seeing nested DualNumber tests in at least divgrad_unit.C, main_unit.C, and navier_unit.h.  Okay, surely that's enough.

**Comment by lindsayad** [comment]
> But that was all it took! When the input type defined value_type but also was convertible to T (because it was a T, in this case, IIRC), then it qualified for both the first constructor and for the second.

Oh I've got you...you were constructing a `DualNumber<ShadowNumber...>` from a `ShadowNumber`...wait that would have been disqualified because `ShadowNumber` doesn't define a `derivatives_type`. But yea I suppose you could concoct some scenario with ambiguity

**Pull request merged**
